backend/services/implementations/notificationService.ts

```diff
@@ -0,0 +1,151 @@
+          link: notification.link,
+        };
+      });
+
+      totalNumberOfNotifications = await MgNotification.where({
+        user,
+      }).countDocuments();
+
+      numberOfUnseenNotifications = await MgNotification.where({
+        user,
+        read: false,
+      }).countDocuments();
+    } catch (error) {
+      Logger.error(
+        `Failed to retrieve notifications. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return {
+      notifications,
+      totalNumberOfNotifications,
+      numberOfUnseenNotifications,
+    };
+  }
+
+  async createNotification(
+    notification: CreateNotificationDTO,
+  ): Promise<NotificationDTO> {
+    let newNotification: Notification;
+    try {
+      newNotification = await MgNotification.create({
+        user: notification.user,
+        message: notification.message,
+        link: notification.link,
+      });
+    } catch (error) {
+      Logger.error(
+        `Failed to create notification. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return {
+      id: newNotification.id,
+      user: newNotification.user,
+      message: newNotification.message,
+      read: newNotification.read,
+      createdAt: newNotification.createdAt,
+      link: newNotification.link,
+    };
+  }
+
+  async updateNotification(
+    notificationId: string,
+    updateNotification: UpdateNotificationDTO,
+  ): Promise<NotificationDTO> {
+    let oldNotification: Notification | null;
+    try {
+      oldNotification = await MgNotification.findByIdAndUpdate(
+        notificationId,
+        updateNotification,
+        { runValidators: true },
+      );
+      if (!oldNotification) {
+        throw new Error(`Notification ${notificationId} not found.`);
+      }
+    } catch (error) {
+      Logger.error(
+        `Failed to update notification. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+
+    // this is digustingly bad ngl
+    const message = updateNotification.message
+      ? updateNotification.message
+      : oldNotification.message;
+    const read = updateNotification.read
+      ? updateNotification.read
+      : oldNotification.read;
+    const link = updateNotification.link
+      ? updateNotification.link
+      : oldNotification.link;
+    return {
+      id: oldNotification.id,
+      user: oldNotification.user,
+      createdAt: oldNotification.createdAt,
+      message,
+      read,
+      link,
+    };
+  }
+
+  async markReadNotifications(userId: string): Promise<UpdateWriteOpResult> {
+    let updates: UpdateWriteOpResult | null;
+    try {
+      updates = await MgNotification.updateMany(
+        { user: userId, read: false },
+        {
+          read: true,
+        },
+      );
+    } catch (error) {
+      Logger.error(
+        `Failed to update notifications. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return updates;
```

Comment:
does this return the updated documents? or the documents as they were found, before the updates?

Reply:
I think all 3 actually + 1 other field